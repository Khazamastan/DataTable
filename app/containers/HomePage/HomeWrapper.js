import styled from 'styled-components';
const HomeWrapper = styled.div`
  margin: 0 auto;
  min-height: 101vh;
  padding: 0 100px 40px 100px;
  background: transparent url(${props => props.bg}) center/cover;
  transition: background 0.3s ease-in;
  .table-container {
    min-height: 400px;
    position: relative;
    .table-loader {
      position: absolute;
      top: 50%;
      left: 0;
      right: 0;
    }
  }
`;

export default HomeWrapper;
