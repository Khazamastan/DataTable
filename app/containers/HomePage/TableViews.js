import styled from 'styled-components';
import React from 'react';

const TitleWrrapper = styled.p`
  span {
    font-weight: 500;
    color: #2196f3;
  }
  &:first-letter {
    text-transform: capitalize;
  }
`;

const ThumbnailCellWrarpper = styled.img`
  padding: 2px;
  max-height: 50px;
  border: 1px solid #ececec;
`;


const LinkCellWrarpper = styled.a`
    color: #2196f3;
    text-decoration: none;
    font-size: 12px;
`;

export const TitleCell = props => {
  const { column, value } = props;
  return (
    <TitleWrrapper>
      <span>{column && column != 'NULL' ? value : '-'}</span>
    </TitleWrrapper>
  );
};

export const ThumbnailCell = props => {
  const { column, value } = props;
  return <ThumbnailCellWrarpper className="img" src={value} />;
};


export const LinkCell = props => {
    const { column, value } = props;
    return <LinkCellWrarpper>{value}</LinkCellWrarpper>;
}