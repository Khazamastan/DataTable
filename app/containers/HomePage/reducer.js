/*
 * SongsReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { LOAD_SONGS_SUCCESS, LOAD_SONGS, LOAD_SONGS_ERROR } from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  currentUser: false,
  songs: false,
};

/* eslint-disable default-case, no-param-reassign */
const photosReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_SONGS:
        draft.loading = true;
        draft.error = false;
        draft.songs = false;
        break;

      case LOAD_SONGS_SUCCESS:
        draft.songs = action.songs;
        draft.loading = false;
        break;

      case LOAD_SONGS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default photosReducer;
