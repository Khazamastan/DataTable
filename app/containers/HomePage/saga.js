/**
 * Gets the songs of the user from Github
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import request from 'utils/request';
import { LOAD_SONGS } from './constants';
import { songsLoaded, songsLoadingError } from './actions';

/**
 * Photos request/response handler
 */
export function* getPhotos() {
  const requestURL = `https://jsonplaceholder.typicode.com/photos`;

  try {
    // Call our request helper (see 'utils/request')
    const songs = yield call(request, requestURL);
    yield put(songsLoaded(songs));
  } catch (err) {
    yield put(songsLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* photosData() {
  // Watches for LOAD_SONGS actions and calls getPhotos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_SONGS, getPhotos);
}
