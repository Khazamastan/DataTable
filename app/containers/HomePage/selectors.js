/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectSongs = state => state.songs || initialState;

const selectRouter = state => state.router;

const makeSelectSongsLoading = () =>
  createSelector(
    selectSongs,
    songsState => songsState.loading,
  );

const makeSelectSongsError = () =>
  createSelector(
    selectSongs,
    songsState => songsState.error,
  );

const makeSelectTotal = () =>
  createSelector(
    selectSongs,
    songsState => songsState.total,
  );

const makeSelectSongs = () =>
  createSelector(
    selectSongs,
    songsState => songsState.songs,
  );

const makeSelectQuery = () =>
  createSelector(
    selectSongs,
    songsState => songsState.query,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectSongs,
  makeSelectSongsLoading,
  makeSelectSongsError,
  makeSelectTotal,
  makeSelectSongs,
  makeSelectLocation,
  makeSelectQuery,
};
