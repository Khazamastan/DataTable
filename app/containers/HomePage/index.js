/* eslint-disable react/prop-types */
/* eslint-disable no-console */
/* eslint-disable no-unused-expressions */
/* eslint-disable no-unused-vars */
/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';

import DataTable from 'components/DataTable';
import LoadingIndicator from 'components/LoadingIndicator';
import {
  makeSelectSongs,
  makeSelectSongsLoading,
  makeSelectSongsError,
} from './selectors';
import { TitleCell, ThumbnailCell, LinkCell } from './TableViews';
import HomeWrapper from './HomeWrapper';

import Section from './Section';
import messages from './messages';
import { fetchSongs } from './actions';
import reducer from './reducer';
import saga from './saga';
import 'react-virtualized/styles.css';
const key = 'home';

export function HomePage({ loading, error, photos, fetchSongsData }) {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });
  useEffect(() => {
    !loading && !photos && fetchSongsData();
  }, []);

  const columns = [
    {
      key: 'checkbox',
      label: '',
      width: '40px',
    },
    {
      key: 'thumbnailUrl',
      label: '',
      width: '60px',
      view: ThumbnailCell,
    },
    {
      key: 'title',
      label: 'Title',
      width: '35%',
      view: TitleCell,
      sort: 1,
    },
    {
      key: 'albumId',
      label: 'Album ID',
      width: '20%',
      numeric: true,
      sort: 1,
    },
    {
      key: 'url',
      label: 'URL',
      view: LinkCell,
      sort: 1,
    },
  ];

  const onRowClickHandler = rowData => {
    console.log(rowData);
  };
  const onSelectRowHandler = (selectedRowIds, selectedRowsOriginal) => {
    console.log(selectedRowIds, selectedRowsOriginal);
  };

  return (
    <article>
      <HomeWrapper>
        <Helmet>
          <meta name="description" content="A Table application homepage" />
        </Helmet>
        <div>
          <Section>
            <div>
              <h2>
                <FormattedMessage {...messages.homeHeader} />
              </h2>
            </div>
            <div className="table-container">
              {!loading && photos && photos.length ? (
                <DataTable
                  data={photos}
                  columns={columns}
                  onRowClick={onRowClickHandler}
                  onSelectRow={onSelectRowHandler}
                />
              ) : (
                <div className="table-loader">
                  <LoadingIndicator />
                </div>
              )}
            </div>
          </Section>
        </div>
      </HomeWrapper>
    </article>
  );
}

HomePage.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  photos: PropTypes.oneOfType([PropTypes.array, PropTypes.bool]),
};

const mapStateToProps = createStructuredSelector({
  photos: makeSelectSongs(),
  loading: makeSelectSongsLoading(),
  error: makeSelectSongsError(),
});

export function mapDispatchToProps(dispatch) {
  return {
    fetchSongsData: () => {
      dispatch(fetchSongs());
    },
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(HomePage);
