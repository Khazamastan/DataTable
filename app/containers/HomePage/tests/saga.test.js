/**
 * Tests for HomePage sagas
 */

import { takeLatest } from 'redux-saga/effects';

import { LOAD_SONGS } from 'containers/HomePage/constants';
// import { songsLoaded, songsLoadingError } from 'containers/HomePage/actions';

import photosData, { getPhotos } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('getPhotos Saga', () => {
  let getPhotosGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getPhotosGenerator = getPhotos();

    const selectDescriptor = getPhotosGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getPhotosGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  // it('should dispatch the songsLoaded action if it requests the data successfully', () => {
  //   const response = [];
  //   const putDescriptor = getPhotosGenerator.next(response).value;
  //   expect(putDescriptor).toEqual(put(songsLoaded(response)));
  // });

  // it('should call the songsLoadingError action if the response errors', () => {
  //   const response = new Error('Some error');
  //   const putDescriptor = getPhotosGenerator.throw(response).value;
  //   expect(putDescriptor).toEqual(put(songsLoadingError(response)));
  // });
});

describe('photosDataSaga Saga', () => {
  const photosDataSaga = photosData();

  it('should start task to watch for LOAD_SONGS action', () => {
    const takeLatestDescriptor = photosDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_SONGS, getPhotos));
  });
});
