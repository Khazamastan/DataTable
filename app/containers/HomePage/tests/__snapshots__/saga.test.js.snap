// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPhotos Saga should call the songsLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://jsonplaceholder.typicode.com/photos",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getPhotos Saga should call the songsLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "songs": undefined,
      "type": "HOME/LOAD_SONGS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getPhotos Saga should dispatch the songsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://jsonplaceholder.typicode.com/photos",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getPhotos Saga should dispatch the songsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "songs": undefined,
      "type": "HOME/LOAD_SONGS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
